Implement invoke defaults for popover

lukewarlow opened the pull request:
#### 8b3a40134afa148adf04eeae3ca908ce0f5a1c40
<pre>
Implement invoke defaults for popover
<a href="https://bugs.webkit.org/show_bug.cgi?id=263513">https://bugs.webkit.org/show_bug.cgi?id=263513</a>

Reviewed by Darin Adler.

This adds the `handleInvokeInternal` logic to toggle popovers.

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/invokers/invoketarget-on-popover-behavior.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/invokers/invoketarget-on-popover-behavior.tentative.html:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/html/semantics/invokers/invoketarget-on-popover-behavior.tentative-expected.txt:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::handlePopoverLightDismiss):
* Source/WebCore/dom/Element.h:
(WebCore::Element::handleInvokeInternal):
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::HTMLElement::handleInvokeInternal):
* Source/WebCore/html/HTMLElement.h:
* Source/WebCore/html/HTMLFormControlElement.cpp:
(WebCore::HTMLFormControlElement::handleInvokeAction):

Canonical link: <a href="https://commits.webkit.org/276694@main">https://commits.webkit.org/276694@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/791f253434d2a079aa4a3f999fb9e1faf761269d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/45354 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/24476 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/47881 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/48018 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/41362 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/47661 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/28698 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/21871 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/37195 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/45932 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/21548 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/39126 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/18297 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/18953 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/40210 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/3401 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/41640 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/40529 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/49736 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/20337 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/16861 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/44243 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/21644 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/39287 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/43055 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10093 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/22004 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/21332 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

lukewarlow commented:
TODO see what the iOS bots come back with and update accordingly

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/f7c862328b9d6d326ac35ac4c6c4ae5809a0d9a6. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/25553

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/f0f48e3a581281e383aa7452ab7566f692d9e7e3. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/25553

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/46399415869c5331d0154675e7869b4c9bdc7932. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/25553

darinadler commented on Source/WebCore/dom/Element.h:
```diff
@@ -620,7 +621,7 @@ class Element : public ContainerNode {
     bool isPopoverShowing() const;
 
-    virtual void handleInvokeInternal(const AtomString&) { }
+    virtual bool handleInvokeInternal(const HTMLFormControlElement*, const AtomString&) { return false; }
```
Can we add an argument name here for the `AtomString`? Not obvious what it was/is for.

Why does this function have a return value? The caller doesn’t look at the return value.

lukewarlow replied:
The return value will be used by future patches. It's used by overrides to check if they should early return. For example dialog will implement this function and call the parents. If it's true (e.g. a popover action occured) it knows it doesn't have any work to do and can just early return.

darinadler commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -1578,6 +1578,49 @@ void HTMLElement::popoverAttributeChanged(const AtomString& value)
+    if (popoverState() == PopoverState::None)
+        return false;
+
+    bool shouldShowPopover = equalIgnoringASCIICase(action, autoAtom())
+        || equalIgnoringASCIICase(action, togglePopoverAtom())
+        || equalIgnoringASCIICase(action, showPopoverAtom());
+    bool shouldHidePopover = equalIgnoringASCIICase(action, autoAtom())
+        || equalIgnoringASCIICase(action, togglePopoverAtom())
+        || equalIgnoringASCIICase(action, hidePopoverAtom());
```
Calling `equalIgnoringASCIICase` does not require or benefit from `AtomString`, so the global storage above is wasted. It would be better to have `ASCIILiteral` constants, which would not require `MainThreadNeverDestroyed` and so be faster and use less storage, and use `equalLettersIgnoringASCIICase`, which is faster. It’s OK to use `AtomString` when we want to do a case sensitive comparison, because that gets optimized to a pointer compare.

We should only evaluate one of these booleans, rather than always computing both but only using one. Something like this:
```
if (isPopoverShowing()) {
    bool shouldHide = equalLettersIgnoringASCIICase …
    if (shouldHide) {
        hidePopover();
        return true;
    }
} else {
    bool shouldShow = equalLettersIgnoringASCIICase …
    if (shouldShow) {
        showPopover(invoker);
        return true;
    }
}
```

darinadler commented on Source/WebCore/html/HTMLElement.h:
```diff
@@ -162,6 +162,8 @@ class HTMLElement : public StyledElement {
     void setPopover(const AtomString& value) { setAttributeWithoutSynchronization(HTMLNames::popoverAttr, value); };
     void popoverAttributeChanged(const AtomString& value);
 
+    bool handleInvokeInternal(const HTMLFormControlElement* invoker, const AtomString&) override;
```
The `invoker` argument should be a reference rather than a pointer since we will never get `nullptr`.

Should use `final` instead of `override`.

lukewarlow replied:
Ah yeah that makes sense, the reason it's a override is because it will be overriden in future, but I'll correct it to final here.

darinadler commented on Source/WebCore/html/HTMLFormControlElement.cpp:
The code in `HTMLFormControlElement::invokeAction` uses a poor idiom:
```
if (!value || value.isNull() || value.isEmpty())
    return autoAtom();
```
It includes two checks for null that do the same thing, and a check for empty, which is true for null also. So is a better way to write code that would do the same thing more efficiently.
```
if (value.isEmpty())
    return autoAtom();
```
or
```
auto& action = attributeWithoutSynchronization(HTMLNames::invokeactionAttr);
return action.isEmpty() ? autoAtom() : action;
```

lukewarlow replied:
That codes been removed as of https://github.com/WebKit/WebKit/pull/25693

darinadler commented on Source/WebCore/html/HTMLFormControlElement.cpp:
```diff
@@ -459,7 +459,7 @@ void HTMLFormControlElement::handleInvokeAction()
 
     if (!event->defaultPrevented())
-        invokee->handleInvokeInternal(action);
+        invokee->handleInvokeInternal(this, action);
```
Why does this code call `invokeAction` **before** dispatching the event? Do we have test cases that cover what happens if the event handler changes the `invokeaction` value? We should. This is the kind of thing that can lead to a non-obvious source of incompatibility between different web browsers when people write strange code.

lukewarlow replied:
It has to call invokeAction before dispatching the event because in a follow up it will only fire the event depending on if the action name meets certain restrictions. It's a good point that we should have a test to ensure that changing the attribute during the event doesn't change what happens.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/26ecdbb74de643e202b088fec457504a48667869. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/25553

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/e5529471ffebfc32ba51c342c802f7e8acb30440. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/25553

darinadler commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -1573,6 +1573,47 @@ void HTMLElement::popoverAttributeChanged(const AtomString& value)
         ensurePopoverData().setPopoverState(newPopoverState);
 }
 
+static const ASCIILiteral togglePopoverLiteral()
```
These can just be:
```
constexpr ASCIILiteral togglePopover = "togglepopover"_s;
```
Really don’t need functions.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/eeb4e7ff07d2cd3eca2aa7f19b7867b6e49f2adf. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/25553

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/16d1f7b54fd109094369b5c434b06e608eb52364. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/25553

lukewarlow committed:
Implement invoke defaults for popover
https://bugs.webkit.org/show_bug.cgi?id=263513

Reviewed by Darin Adler.

This adds the `handleInvokeInternal` logic to toggle popovers.

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/invokers/invoketarget-on-popover-behavior.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/invokers/invoketarget-on-popover-behavior.tentative.html:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/html/semantics/invokers/invoketarget-on-popover-behavior.tentative-expected.txt:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::handlePopoverLightDismiss):
* Source/WebCore/dom/Element.h:
(WebCore::Element::handleInvokeInternal):
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::HTMLElement::handleInvokeInternal):
* Source/WebCore/html/HTMLElement.h:
* Source/WebCore/html/HTMLFormControlElement.cpp:
(WebCore::HTMLFormControlElement::handleInvokeAction):

Canonical link: https://commits.webkit.org/276694@main

webkit-commit-queue commented:
Committed 276694@main (8b3a40134afa): <https://commits.webkit.org/276694@main>

Reviewed commits have been landed. Closing PR #25553 and removing active labels.
